pkg/deployer/lib/controller.go

```diff
@@ -144,60 +147,160 @@ func NewController(log logr.Logger,
+		}
 	}
 	logger.Info("reconcile deploy item")
+	hookRes = extension.AggregateHookResults(hookRes, tmpHookRes)
+	// reset AbortReconcile, since it could be 'true' at this point, which would wrongly cause an abort after the next hook
+	hookRes.AbortReconcile = false
 
-	errHdl := HandleErrorFunc(logger, c.lsClient, c.lsEventRecorder, deployItem)
+	tmpHookRes, err = c.deployer.ExtensionHooks().ExecuteHooks(ctx, extensionLogger, di, target, extension.BeforeAnyReconcileHook)
+	if err != nil {
+		return reconcile.Result{}, errHdl(ctx, err)
+	}
+	hookRes = extension.AggregateHookResults(hookRes, tmpHookRes)
+	if hookRes.AbortReconcile {
+		return returnAndLogReconcileResult(logger, *hookRes), nil
+	}
 
-	switch lsv1alpha1.Operation(lsv1alpha1helper.GetOperation(deployItem.ObjectMeta)) {
+	switch lsv1alpha1.Operation(lsv1alpha1helper.GetOperation(di.ObjectMeta)) {
 	case lsv1alpha1.AbortOperation:
-		if err := errHdl(ctx, c.deployer.Abort(ctx, deployItem, target)); err != nil {
+		logger.V(5).Info("handle deploy item abort")
+		tmpHookRes, err = c.deployer.ExtensionHooks().ExecuteHooks(ctx, extensionLogger, di, target, extension.BeforeAbortHook)
+		if err != nil {
+			return reconcile.Result{}, errHdl(ctx, err)
+		}
+		hookRes = extension.AggregateHookResults(hookRes, tmpHookRes)
+		if hookRes.AbortReconcile {
+			return returnAndLogReconcileResult(logger, *hookRes), nil
+		}
+		if err := errHdl(ctx, c.deployer.Abort(ctx, di, target)); err != nil {
 			return reconcile.Result{}, err
 		}
 	case lsv1alpha1.ForceReconcileOperation:
-		if err := errHdl(ctx, c.deployer.ForceReconcile(ctx, deployItem, target)); err != nil {
+		logger.V(5).Info("handle deploy item force-reconcile")
+		tmpHookRes, err = c.deployer.ExtensionHooks().ExecuteHooks(ctx, extensionLogger, di, target, extension.BeforeForceReconcileHook)
+		if err != nil {
+			return reconcile.Result{}, errHdl(ctx, err)
+		}
+		hookRes = extension.AggregateHookResults(hookRes, tmpHookRes)
+		if hookRes.AbortReconcile {
+			return returnAndLogReconcileResult(logger, *hookRes), nil
+		}
+		if err := errHdl(ctx, c.deployer.ForceReconcile(ctx, di, target)); err != nil {
 			return reconcile.Result{}, err
 		}
 	default:
 
-		if !deployItem.DeletionTimestamp.IsZero() {
-			if err := errHdl(ctx, c.delete(ctx, deployItem, target)); err != nil {
+		if !di.DeletionTimestamp.IsZero() {
+			logger.V(5).Info("handle deploy item deletion")
+			tmpHookRes, err = c.deployer.ExtensionHooks().ExecuteHooks(ctx, extensionLogger, di, target, extension.BeforeDeleteHook)
+			if err != nil {
+				return reconcile.Result{}, errHdl(ctx, err)
+			}
+			hookRes = extension.AggregateHookResults(hookRes, tmpHookRes)
+			if hookRes.AbortReconcile {
+				return returnAndLogReconcileResult(logger, *hookRes), nil
+			}
+			if err := errHdl(ctx, c.delete(ctx, di, target)); err != nil {
+				return reconcile.Result{}, err
+			}
+		} else {
+			// default reconcile
+			logger.V(7).Info("handle deploy item reconcile")
```

Comment:
It's intended. All the other cases are 'more interesting' than a standard reconcile, but I wanted to have a log output there too, if one wants to ensure that this case is actually executed while debugging. To not spam the logs, I decided for the higher verbosity here. Do you think it would be better to use `5` here too?

Reply:
sounds reasonable